docs/architecture/adr-061-inter-validator-set-messaging.md

```diff
@@ -0,0 +1,108 @@
+# ADR 061: Inter validator set messaging
+
+## Changelog
+
+- 2021-09-29: Initial version of the document
+
+## Context
+
+### Definitions
+
+For the needs of this document, we define the following types of nodes:
+
+- Validator nodes, which participate in the consensus protocol, verify and sign blocks,
+- Full nodes, which store all blocks from the blockchain.
+
+Note: there are also light nodes, but light nodes are not relevant for the discussion.
+
+Validator Set is a group of Validator nodes responsible for the consensus at a given time. Only one Validator Set can
+be active at the given time. After a predefined number of blocks, the ABCI app initiates the Validator Set rotation to
+make another Validator Set active. 
+
+### Problem statement
+
+The consensus process requires direct communication between validators, as full nodes do not accept nor propagate consensus protocol messages (like votes).
+
+Every node, regardless of its type, selects peer nodes they connect to on a random basis. There is no differentiation
+between full and Validator nodes, and there is no method to ensure that each Validator is (directly or through other
+validators) connected to all other validators. In an extreme case, a validator node can be connected only to full
+nodes. It will block its communication and effectively exclude it from participation in the consensus.
+
+### Solution
+
+Each Validator node shall allocate a preconfigured number of connectivity slots for communication with other Validators
+that belong to the same Validator Set.
+
+ABCI application manages the rotation of Validator Sets. When rotation is needed, the application sends information about the new Validator Set in response to the `EndBlock` message. Each member of the active Validator Set establishes connections with a predefined number of Peer Validators belonging to the same Validator Set. A list of Peer Validators is determined using the algorithm described in [DIP 0006](https://github.com/dashpay/dips/blob/master/dip-0006.md#building-the-set-of-deterministic-connections).
+
+## Alternative Approaches
+
+> TODO: This section contains information around alternative options that are considered before making a decision.
+> It should contain a explanation on why the alternative approach(es) were not chosen.
+
+## Decision
+
+> TODO: This section records the decision that was made.
```

Comment:
We could add the discussion that we had on another solution 1 or 3, to send the whole masternode list and quorum list to Tenderdash and potentially rotate on its side. And we agree that probably better to keep the current Tendermint design and concept being a logic-less consensus library and keep all business logic in the ABCI application. Also, there is no point to optimize data sending between ABCI app and Tenderdash since we use unixsocket and potentially compile both parts together.

Reply:
I did add what I understand. I don't understand the scenario 3, if you can clarify / describe, it would be nice.

Reply:
Scenario one was to still trigger rotation from ABCI app passing quorum id to rotate. Scenario three was that Tendermint itself decides on rotation if I don't mixed up everything :)